### ISSUE TITLE
JSON3 hangs IE7 with particular json data with long string property values.

### ISSUE THREAD

**Issue by jaxley** [bestiejs/json3#23]
I created a jsperf to demonstrate the issue.  Even in IE8 the issue is pronounced (allowing a mere 6 ops/second).  http://jsperf.com/json-parser-perf-ie-largestringproperties

We experienced an issue where json3 was causing the IE7 browser to literally hang.  I've tried running the jsperf in IE7 and it has only hung the browser thus far so there is no data yet.  IE8 took a while but it did finish.

Without IE profiler, I can see if I can generate some metrics from dynatrace to identify which specific function(s) or aspects of the functions might be the culprits.  

EDIT: Dynatrace shows only one stand-out function and that is the lex() function when processing the string property value itself.  It took 4779ms to parse that one string.  I'm not sure why sometimes the browser hangs in our real environment (or the jsperf) but it doesn't hang when running the same parse in isolation, other than for the 5 seconds or more it's parsing and the cpu is pegged.

``` xml
<?xml version="1.0" encoding="utf-8"?><browser_purepath_tree><nodeinfo level="7" function="lex()" arguments="" return="@&lt;div class=&quot;columns columns-aligned ...TRUNCATED for brevity" start_ms="7658.334229629743" total_ms="4778.469051234164" exec_ms="4778.46923828125" exec="38.39862" api="json3, JavaScript" agent="Browser"/></browser_purepath_tree>
```

**Comment by oxyc** [comment]
The main issue is commit cefa14ad99b

```diff
-          Source = "" + source;
+          Source = new String(source);
```

v8 bailsout on charCodeAt as it's an object instead of a string. See [code-stubs.h](https://github.com/v8/v8/blob/aec5abab1e9faf1ab0949d2f068093544ba2bc40/src/code-stubs.h#L1683:L1686).

**Comment by oxyc** [comment]
I ran a quick proof-of-concept with simply validating the json with the current recursive descent parser, and then using eval. Speed does improve significantly on newer browsers. Unfortunately IE9 turned out to be a lot slower, and I do not have access to IE7-8, so it requires more research.

http://jsperf.com/json-parser-perf-ie-largestringproperties/18

The JSON3_EVAL is the latest version with the diff https://gist.github.com/oxyc/6d02b15b745638805afd (tests pass)

**Comment by d10** [comment]
@oxyc Thanks for the research, I'll remove the `new String` usage. Will experiment with the others.

**Comment by d10** [comment]
Here's the latest jsPerf: http://jsperf.com/json-parser-perf-ie-largestringproperties/19.

**Comment by d10** [comment]
Also, I'm not too sure, but I feel that one of the main reasons JSON3.parse is slow is due to the number of function calls (`get` -> `lex` -> ...). If there was a non-recursive way of parsing, I'm sure that would speed things up.

**Comment by oxyc** [comment]
For reference, I was curious to what it would look like if string literals were validated with a regex instead. This increased the speed in chrome (past nativeJSON for this particular edge case) but decreased in IE9 once again.

I basically replaced the string literal loop in `lex()` with a naive regex (fails some tests but works well enough for benchmarking)

```js
                var reNeedsEscape = /"([^"\\]*|\\["\\bfnrt\/]|\\u[0-9a-fA-F]{4})*"/;
                var reEscape = RegExp(reNeedsEscape.source, "g");
                var res = reEscape.exec(source.slice(Index));
                if (res) Index += res[0].length;
                else abort();
                return "@"
```

In case you want to check IE8 here's the jsPerf: http://jsperf.com/json-parser-perf-ie-largestringproperties/20 (includes the eval parsing modification from earlier).

**Comment by d10** [comment]
I think that it’s not very important to have better performance for modern browsers (Chrome) as they would already have valid `JSON.parse` implementations. We should instead focus on older browsers (IE 6 - 8, etc.).

**Comment by oxyc** [comment]
Modern browsers could be valid for mobile.

**Comment by d10** [comment]
Yep, I’m also experimenting, currently with a non-recursive implementation. Also, if you need access to IE 8, you could use an online service like [Sauce Labs](https://saucelabs.com/).

**Comment by d10** [comment]
Here’s my non-recursive implementation: https://github.com/d10/json3/commit/507278b9f9684b608bec3a2f2a626a9d44797a87. I haven’t tested it yet, will try to do so soon.

**Comment by kitcambridge** [comment]
@d10 Nice! Another approach might be to use a validating parser (here's a [very rough, untested port](https://gist.github.com/kitcambridge/b5365e93527b2282fcc8) of Go's `encoding/json` state machine), followed by `new Function('return (' + source + )')`. That has its own drawbacks, though. I'll run some benchmarks once I have access to my VMs again.

**Comment by d10** [comment]
Here’s a jsPerf: http://jsperf.com/json-parser-perf-ie-largestringproperties/23. Seems much slower than current. @oxyc’s regex implementation seems to be the fastest. I will also try to experiment with my non recursive implementation, which currently dies not work.

**Comment by kitcambridge** [comment]
Cool; RegExp it is. We can tighten it to perform additional validation and measure its impact. Thanks, @oxyc!
